File: src/port-forwarder.js
const net = require('net');
const { v4: uuidv4 } = require('uuid');
const readline = require('readline');
const {
  getKernelSpecs,
  startNewKernel,
  connectToKernelWebSocket,
  executeCode,
  setupPortForwarderInKernel,
  shutdownKernel,
} = require('./kernelManager');
require('dotenv').config();
// --- Helper function for command-line prompts ---
function promptForVar(question, defaultValue) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  return new Promise(resolve => {
    rl.question(`${question} [${defaultValue}]: `, answer => {
      rl.close();
      resolve(answer || defaultValue);
    });
  });
}

// --- Declare key variables in a higher scope for access in the shutdown handler ---
let ws = null;
let kernel = null;
let server = null;
let JUPYTER_PROXY_URL = process.env.JUPYTER_PROXY_URL;
const clientSockets = new Map();


(async () => {
  try {
    // --- 1. Prompt for missing configuration ---
    console.log('--- Port Forwarder Configuration ---');
    if (!JUPYTER_PROXY_URL) {
      JUPYTER_PROXY_URL = await promptForVar('Enter Jupyter Proxy URL', 'http://127.0.0.1:8888');
    }
    console.log(process.env.LOCAL_PORT)
    const LOCAL_PORT = process.env.LOCAL_PORT || await promptForVar('Enter Local Port to listen on', '9000');
    const REMOTE_HOST = process.env.REMOTE_HOST || await promptForVar('Enter Remote Host to connect to (from the kernel)', '127.0.0.1');
    const REMOTE_PORT = process.env.REMOTE_PORT || await promptForVar('Enter Remote Port to connect to', '8000');
    console.log('------------------------------------');


    // --- 2. Connect to the Jupyter Kernel ---
    console.log('Connecting to Jupyter Kernel...');
    const specs = await getKernelSpecs(JUPYTER_PROXY_URL);
    kernel = await startNewKernel(JUPYTER_PROXY_URL, specs.default);
    ws = await connectToKernelWebSocket(JUPYTER_PROXY_URL, kernel.id);
    console.log(`Connected to Kernel ID: ${kernel.id}`);


    // --- 3. Inject the Python port forwarding logic into the kernel ---
    await setupPortForwarderInKernel(ws, REMOTE_HOST, REMOTE_PORT);


    // --- 4. Listen for messages coming BACK from the kernel ---
    ws.on('message', raw => {
      const outer = JSON.parse(raw.toString());

      // Handle our port-forwarder prints (may contain several JSON objects concatenated)
      if (
        outer.msg_type === 'stream' &&
        outer.content?.name === 'stdout' &&
        outer.content.text.includes('FORWARDER_MSG:')
      ) {
        const segments = outer.content.text
          .split('\n')                       // each Python `print` ends with \n
          .filter(Boolean)                   // drop empty lines
          .filter(l => l.startsWith('FORWARDER_MSG:'));

        for (const line of segments) {
          try {
            const payload = JSON.parse(
              line.slice('FORWARDER_MSG:'.length)  // strip prefix
            );
            handleForwarder(payload);
          } catch (e) {
            console.error('Bad JSON from kernel:', e, line);
          }
        }
        return;                               // nothing more to do for this outer msg
      }

      if (outer.msg_type === 'error') {
        console.error('--- KERNEL ERROR ---');
        console.error('Name:', outer.content.ename);
        console.error('Value:', outer.content.evalue);
        console.error('Traceback:\n' + outer.content.traceback.join('\n'));
        console.error('--------------------');
        return;
      }

      if (outer.msg_type === 'stream' && outer.content.name === 'stderr') {
        console.log(`[KERNEL STDERR]: ${outer.content.text.trim()}`);
        return;
      }
    });
    // --- Helper to process each decoded port-forwarder payload ---
    function handleForwarder({ type, conn_id, data, error }) {
      const localSocket = clientSockets.get(conn_id);
      if (!localSocket) return;

      switch (type) {
        case 'data':
          localSocket.write(Buffer.from(data, 'base64'));
          break;
        case 'close':
          console.log(`Connection ${conn_id} closed by remote. Closing local socket.`);
          localSocket.end();
          clientSockets.delete(conn_id);
          break;
        case 'connect_error':
          console.error(`Kernel failed to connect for ${conn_id}: ${error}`);
          localSocket.end();
          clientSockets.delete(conn_id);
          break;
      }
    }


    // --- 5. Create the local TCP server ---
    server = net.createServer(localSocket => {
      // ... (This section is unchanged)
      const conn_id = uuidv4();
      console.log(`New local client connected. Assigning ID: ${conn_id}`);
      clientSockets.set(conn_id, localSocket);
      executeCode(ws, `start_connection("${conn_id}")`);
      localSocket.on('data', chunk => {
        const b64_data = chunk.toString('base64');
        const code = `forward_data("${conn_id}", "${b64_data}")`;
        executeCode(ws, code);
      });
      localSocket.on('close', () => {
        console.log(`Local client ${conn_id} disconnected.`);
        if (clientSockets.has(conn_id)) {
          clientSockets.delete(conn_id);
          executeCode(ws, `close_connection("${conn_id}")`);
        }
      });
      localSocket.on('error', (err) => {
        console.error(`Error on local socket ${conn_id}:`, err);
        if (clientSockets.has(conn_id)) {
          clientSockets.delete(conn_id);
          executeCode(ws, `close_connection("${conn_id}")`);
        }
      });
    });

    server.listen(LOCAL_PORT, () => {
      console.log(`\n✅ TCP Port Forwarder is running.`);
      console.log(`   Forwarding connections from localhost:${LOCAL_PORT} -> (via kernel) -> ${REMOTE_HOST}:${REMOTE_PORT}`);
      console.log(`   Press CTRL+C to shut down gracefully.\n`);
    });

  } catch (e) {
    console.error("\n❌ Fatal initialization error:", e.message);
    // If the error happened after the kernel was created, try to clean it up.
    if (kernel) {
      await shutdownKernel(JUPYTER_PROXY_URL, kernel.id);
    }
    process.exit(1);
  }
})();

// --- 6. Handle graceful shutdown on CTRL+C ---
process.on('SIGINT', async () => {
  console.log('\n\nCaught interrupt signal. Shutting down gracefully...');

  // 1. Stop accepting new connections
  if (server) {
    server.close(() => {
      console.log('Local server closed.');
    });
  }

  // 2. Close all active connections on the kernel side
  if (ws && clientSockets.size > 0) {
    console.log(`Closing ${clientSockets.size} active connection(s) in kernel...`);
    for (const conn_id of clientSockets.keys()) {
      executeCode(ws, `close_connection("${conn_id}")`);
    }
  }

  // 3. Destroy all local sockets
  for (const socket of clientSockets.values()) {
    socket.destroy();
  }
  clientSockets.clear();

  // 4. Shut down the remote kernel itself
  if (kernel) {
    await shutdownKernel(JUPYTER_PROXY_URL, kernel.id);
  }

  // Give a moment for cleanup messages to be processed
  setTimeout(() => {
    console.log('Cleanup complete. Exiting.');
    process.exit(0);
  }, 1000);
});

File: src/port_forwarder.py
import socket
import threading
import json
import base64
import sys
import time

# These will be replaced by kernelManager.js before execution
REMOTE_HOST = "{{REMOTE_HOST}}"
REMOTE_PORT = int("{{REMOTE_PORT}}")

# A thread-safe dictionary to hold connections
connections = {}
lock = threading.Lock()

def _log(message):
    """Helper for stderr logging to not interfere with stdout data channel."""
    print(message, file=sys.stderr, flush=True)

def _send_to_node(message):
    """
    Safely prints a message to stdout for Node.js to capture.
    The prefix is crucial for distinguishing our messages.
    """
    print(f"FORWARDER_MSG:{json.dumps(message)}", flush=True)


def _listen_for_remote_data(conn_id, sock):
    """
    This function runs in a dedicated thread for each connection.
    It reads data from the remote service and forwards it to Node.js.
    """
    _log(f"Listener thread started for conn_id: {conn_id}")
    try:
        while True:
            # Reading in chunks
            data = sock.recv(4096)
            if not data:
                # The remote server closed the connection
                _log(f"Remote connection closed for conn_id: {conn_id}")
                break

            # Base64 encode to safely transmit binary data via JSON
            b64_data = base64.b64encode(data).decode('utf-8')
            
            response = {
                "type": "data",
                "conn_id": conn_id,
                "data": b64_data
            }
            _send_to_node(response)

    except Exception as e:
        _log(f"Error in listener for {conn_id}: {e}")
    finally:
        # Notify Node.js that this connection is now closed from the remote side
        _send_to_node({"type": "close", "conn_id": conn_id})
        close_connection(conn_id)


def start_connection(conn_id):
    """
    Called by Node.js to initiate a new TCP connection to the remote target.
    """
    _log(f"Attempting to start connection {conn_id} to {REMOTE_HOST}:{REMOTE_PORT}")
    try:
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.connect((REMOTE_HOST, REMOTE_PORT))
        
        with lock:
            connections[conn_id] = sock

        # Start a listener thread to read from this new socket
        thread = threading.Thread(target=_listen_for_remote_data, args=(conn_id, sock))
        thread.daemon = True # Allows main program to exit even if threads are running
        thread.start()
        
        _log(f"Successfully connected and started listener for {conn_id}")
        _send_to_node({"type": "connect_success", "conn_id": conn_id})

    except Exception as e:
        _log(f"Failed to connect for {conn_id}: {e}")
        _send_to_node({"type": "connect_error", "conn_id": conn_id, "error": str(e)})


def forward_data(conn_id, b64_data):
    """
    Called by Node.js to forward data from the local client to the remote service.
    """
    with lock:
        sock = connections.get(conn_id)
    
    if sock:
        try:
            # Decode the data and send it
            data = base64.b64decode(b64_data)
            sock.sendall(data)
        except Exception as e:
            _log(f"Error forwarding data for {conn_id}: {e}")
            # This connection is likely broken, so clean it up
            close_connection(conn_id)
    else:
        _log(f"Could not forward data: No active socket for conn_id {conn_id}")


def close_connection(conn_id):
    """
    Called by Node.js (or internally) to close a specific connection.
    """
    with lock:
        sock = connections.pop(conn_id, None)

    if sock:
        try:
            _log(f"Closing connection for {conn_id}")
            sock.close()
        except Exception as e:
            _log(f"Error closing socket for {conn_id}: {e}")

_log("Python port forwarder functions defined and ready.")
